### PR TITLE
[`useless_vec`]: lint `vec!` invocations when a slice or an array would do

### PR DESCRIPTION
**Pull request by y21**
First off, sorry for that large diff in tests. *A lot* of tests seem to trigger the lint with this new change, so I decided to `#![allow()]` the lint in the affected tests to make reviewing this easier, and also split the commits up so that the first commit is the actual logic of the lint and the second commit contains all the test changes. The stuff that changed in the tests is mostly just line numbers now. So, as large as the diff looks, it's not actually that bad. 😅
I manually went through all of these to find out about edge cases and decided to put them in `tests/ui/vec.rs`.

For more context, I wrote about the idea of this PR here: https://github.com/rust-lang/rust-clippy/issues/2262#issuecomment-1579155257 (that explains the logic)

Basically, it now also considers the case where a `Vec` is put in a local variable and the user only ever does things with it that one could also do with a slice or an array. This should catch a lot more cases, and (at least from looking at the tests) it does.

changelog: [`useless_vec`]: lint `vec!` invocations when a slice or an array would do (also considering local variables now)

### CODE REVIEW

File: clippy_lints/src/vec.rs [hunk 1]
```diff
@@ -46,16 +51,70 @@ declare_clippy_lint! {
...
+    } else {
+        false
+    }
+}
+
+/// Checks if the given expression is a method call to a `Vec` method
+/// that also exists on slices. If this returns true, it means that
+/// this expression does not actually require a `Vec` and could just work with an array.
+fn is_allowed_vec_method(cx: &LateContext<'_>, e: &Expr<'_>) -> bool {
+    const ALLOWED_METHOD_NAMES: &[&str] = &["len", "as_ptr", "is_empty"];
```
Comment (Centri3) [inline clippy_lints/src/vec.rs:66]: I'm a bit curious why `len` and `is_empty` are allowed.

**Review by Centri3 — COMMENTED** [review]
This looks good, thanks! Just one choice on `ALLOWED_METHOD_NAMES` I'm curious about.

File: clippy_lints/src/vec.rs (hunk 1 above)
Comment (y21) [reply clippy_lints/src/vec.rs:66]: The method names that are checked for here exist on both slices and `Vec<T>`.
So, given a variable `ve` of type `Vec<i32>`, a call expression like `ve.is_empty()` does not actually call `slice::is_empty`, but `Vec::is_empty`, which also means that `adjusts_to_slice(ve)` would return false.

If we didn't have `len` and `is_empty` here, the lint would assume that those are `Vec<T>` methods that really do require a `Vec<T>` (like `push` and `pop`, for example).
For `len` and `is_empty` specifically, that isn't the case. These two methods also exist on slices, and we still want to lint for cases like these, so we allow those:
```rs
let ve = vec![1, 2];
dbg!(ve.is_empty()); // Vec::is_empty, but this would also work fine if ve was `[1, 2]`
```

Comment (Centri3) [reply clippy_lints/src/vec.rs:66]: Ah, I misunderstood then :sweat_smile: I thought those were exceptions. Yeah, that sounds fine then.

File: tests/ui/vec.rs [hunk 2]
```diff
@@ -75,4 +81,44 @@ fn main() {
     for a in vec![1; 201] {
         println!("{:?}", a);
     }
+
+    // Do lint
+    let mut x = vec![1, 2, 3];
+    x.fill(123);
+    dbg!(x[0]);
```
Comment (y21) [inline tests/ui/vec.rs:88]: One interesting case I just thought of is that changing a vec to an array can sometimes lead to compile errors when indexing with a constant that is out of bounds, because the compiler checks index expressions for arrays but not vecs:

```rs
let ve = vec![1, 2];
ve[3]; // ok (no compile error)

let arr = [1, 2];
arr[3]; // compile error
```

Not really sure what to do here. Accept it as a "known issue"? I doubt this is something that people run into by accident?

Comment (Centri3) [reply tests/ui/vec.rs:88]: "Known issue" is likely fine IMO, this would likely be helpful anyway

File: clippy_lints/src/vec.rs (hunk 1 above)
Comment (Manishearth) [reply clippy_lints/src/vec.rs:66]: nit: please add a comment to this const explaining what it's about

File: tests/ui/vec.rs (hunk 2 above)
Comment (Manishearth) [reply tests/ui/vec.rs:88]: works for me

**Review by Manishearth — APPROVED** [review]
